File: 3term/1.2. local network/1.2. local network/Behavior.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace _1._2.local_network
{
    /// <summary>
    /// Class that regiment behavior of system in infection process.
    /// </summary>
    public class Behavior
    {
        private Computer[] computers;
        bool[] infectedComputers;

        /// <summary>
        /// Change a system condition because of thread of infection.
        /// </summary>
        /// <param name="matrix"></param>
        /// <param name="computersFromStuff"></param>
        /// <param name="extraIndex"></param>
        public Behavior(int[,] matrix, Computer[] computersFromStuff)
        {
            computers = computersFromStuff;
            infectedComputers = new bool[matrix.GetLength(0)];
            FillingOfArray(infectedComputers);
        }

        /// <summary>
        /// Fill bool array that demonstrate computers which is infect.
        /// </summary>
        /// <param name="infectedComputers"></param>
        /// <returns></returns>
        public bool[] FillingOfArray(bool[] infectedComputers)
        {
            for (int i = 0; i < infectedComputers.GetLength(0); ++i)
            {
                infectedComputers[i] = computers[i].GetHealthCondition();
            }
            return infectedComputers;
        }

        /// <summary>
        /// Auxiliary process of infection.
        /// </summary>
        /// <param name="infectedComputers"></param>
        /// <param name="matrix"></param>
        /// <param name="extraIndex"></param>
        public void InternalInfection(bool[] infectedComputers, int[,] matrix, int extraIndex)
        {
            infectedComputers = this.FillingOfArray(infectedComputers);

            for (int i = 0; i < infectedComputers.Length; ++i)
            {
                if (infectedComputers[i])
                {
                    for (int j = 0; j < infectedComputers.GetLength(0); ++j)
                    {
                        if ((matrix[i, j] == 1) && ( !((computers[j].GetHealthCondition()) && (computers[i].GetHealthCondition())) ))
                        {
                            computers[j].TryToInfect(extraIndex);
                        }
                    }
                }
            }
            infectedComputers = this.FillingOfArray(infectedComputers);
         }

        /// <summary>
        /// Get array of coputers that are infected.
        /// </summary>
        /// <returns></returns>
        public bool[] GetArray()
        {
            return infectedComputers;
        }

        /// <summary>
        /// Get health condition of intrested computer.
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public bool GetComputerHealthCondition(int name)
        {
            return computers[name].GetHealthCondition();
        }
    }
}

File: 4term/5.1. localNetwork/5.1. localNetwork/Behavior.fs
module Behavior
open Computer
open System
open System.Collections.Generic
open System.Linq
open System.Text
open System.Threading.Tasks
open System.IO
open Microsoft.FSharp.Collections

/// Class that describe behavior of system in infection process.
type Behavior() = 
     let mutable compList : Computer list = []
     let mutable infectedComputers = Array.create 4 false//is private?

/// Change a system condition because of thread of infection.
     member b.Behavior (matrix : int [,]) (computersFromStuff : Computer list) = 
         compList <- computersFromStuff
         infectedComputers <- Array.create (Array2D.length1 matrix) false
         b.FillingOfArray

/// Fill bool list that demonstrate computers which is infect.
     member b.FillingOfArray = 
         for i = 0 to (Array.length infectedComputers - 1) do
             infectedComputers.[i] <- compList.[i].GetHealthCondition
         infectedComputers

/// Auxiliary process of infection.
     member b.InternalInfection (matrix : int[,]) (extraIndex : int) = 
         infectedComputers <- b.FillingOfArray
         for i = 0 to (infectedComputers.Length - 1) do
             if infectedComputers.[i] then
                for j = 0 to (infectedComputers.Length - 1) do
                    let forI = (compList.[i].GetHealthCondition)
                    let forJ = (compList.[j].GetHealthCondition)
                    let compose = forI && forJ
                    let notCompose = not compose
                    let mIJ = matrix.[i, j]
                    if ( (matrix.[i, j] = 1) && ( notCompose ) ) then
                       compList.[j].TryToInfect(extraIndex)
         infectedComputers = b.FillingOfArray
  
/// Get array of coputers that are infected.
     member b.GetArray() = infectedComputers

/// Get health condition of intrested computer.
     member b.GetComputerHealthCondition (name : int) = compList.[name].GetHealthCondition

